feat(ivy): implement listing lazy routes for specific entry point in `ngtsc`

gkalpak opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


## PR Type
- [x] Feature


## What is the current behavior?
`ngtsc`'s `listLazyRoutes()` does not support listing routes for a specific entry point. `@angular/cli` calls this API to get a list of lazy-loaded entry points (routes) so that the bundler can create separate bundles for each. Under certain circumstances (e.g. `jit` mode), the cli may need to list routes for a specific root entry point only.

Jira issue: [FW-860]


## What is the new behavior?
`ngtsc`'s `listLazyRoutes()` supports listing routes for a specific entry point.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


## Other information
Related: angular/angular-cli#13532


[FW-860]: https://angular-team.atlassian.net/browse/FW-860

gkalpak committed:
refactor(ivy): create lazy route keys that are similar to ngtools lazy routes

This will make it easier to retrieve routes for specific entry points in
`listLazyRoutes()` (which is necessary for CLI integration but not yet
implemented).

gkalpak committed:
refactor(ivy): remove unnecessary method from ngtsc's `RouterEntryPoint`

gkalpak committed:
feat(ivy): implement listing lazy routes for specific entry point in `ngtsc`

Related: angular/angular-cli#13532

Jira issue: FW-860

alfaproject commented on packages/compiler-cli/test/ngtools_api_spec.ts:
```diff
@@ -60,7 +59,7 @@ describe('ngtools_api (deprecated)', () => {
   }
 
-  fixmeIvy('FW-629: ngtsc lists lazy routes').it('should list lazy routes recursively', () => {
+  fit('should list lazy routes recursively', () => {
```
Oopsie

gkalpak committed:
test(ivy): fix `ngtools_api` test under ivy

gkalpak committed:
fix(ivy): support listing lazy route for project-root-relative entry point in `ngtsc`

I don't know of any use of this API with a project-root-relative path
(i.e. the cli will always call it with an absolute path), but keeping
the API backwards compatible just in case.

gkalpak commented on packages/compiler-cli/test/ngtools_api_spec.ts:
```diff
@@ -37,7 +38,7 @@ describe('ngtools_api (deprecated)', () => {
         export class ErrorComp2 {}
 
         @NgModule({
-          declarations: [ErrorComp2, NonExistentComp],
```
`ngtsc` complained about non-existent component.

alxhub replied:
As well it should!

petebacondarwin reviewed (approved):
Awesome PR. Some really nice computation.
A few Qs and suggestions but looks good.

Do we need to worry about / test routes that point to files that are in a different source root folder?

petebacondarwin commented on packages/compiler-cli/src/ngtsc/routing/src/analyzer.ts:
```diff
@@ -48,18 +48,53 @@ export class NgModuleRouteAnalyzer {
   }
 
-  listLazyRoutes(): LazyRoute[] {
+  listLazyRoutes(entryModuleKey?: string|undefined): LazyRoute[] {
```
Yeah I know but the end result is the same for the type in the body of the function... if it has a `?` then the value might be `undefined`.

gkalpak replied:
AFAICT, it doesn't make a different for TS, but I was keeping consistent with [program.ts][1] (just in case I miss something).

[1]: https://github.com/angular/angular/blob/80eb608baa5b635c6366c86f197cf9030b124f58/packages/compiler-cli/src/ngtsc/program.ts#L193

alxhub commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -2022,91 +2024,410 @@ describe('ngtsc behavioral tests', () => {
-    expect(routes[0].module.filePath.endsWith('/test.ts')).toBe(true);
-    expect(routes[0].referencedModule.filePath.endsWith('/lazy.ts')).toBe(true);
-  });
+  describe('listLazyRoutes()', () => {
```
I'm actually planning on doing that soon anyway, for all of the tests in it. Stay tuned!

alxhub commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -2022,91 +2024,410 @@ describe('ngtsc behavioral tests', () => {
+                name: toModuleName,
+                filePath: jasmine.stringMatching(toModulePath),
+              }),
+        } as unknown as LazyRoute);
```
Feel free to turn off the clang-formatter and do a better job of making the body readable :)

alxhub commented:
(Requesting that you make the changes Pete suggested!)

gkalpak commented on packages/compiler-cli/test/ngtools_api_spec.ts:
```diff
@@ -19,7 +19,8 @@ describe('ngtools_api (deprecated)', () => {
   function createProgram(rootNames: string[]) {
-    const options = testSupport.createCompilerOptions();
+    const options =
+        testSupport.createCompilerOptions({enableIvy: ivyEnabled && 'ngtsc' as 'ngtsc'});
```
Without it, the type of the object ends up as `{enabledIvy: false | string}`, which TS complains is not assignable to `'ngtsc' | boolean | ...`.

petebacondarwin replied:
Really? That is not what I see. `{enableIvy: ivyEnabled && 'ngtsc'}` appears to be enough.

gkalpak replied:
Oh, you are right. Originally, I had it as a separate variable:

```ts
const overrideOptions = {enableIvy: ivyEnabled && 'ngtsc' as 'ngtsc'};
const options = testSupport.createCompilerOptions(overrideOptions);
```

In that case, TS complains without the `as 'ngtsc'`, but when using the object literal directly, it is not necessary :tada:
Good catch!

gkalpak commented on packages/compiler-cli/test/ngtools_api_spec.ts:
```diff
@@ -62,7 +63,8 @@ describe('ngtools_api (deprecated)', () => {
 
   fixmeIvy('FW-629: ngtsc lists lazy routes').it('should list lazy routes recursively', () => {
```
I removed it is a subsequent commit, because this commit (although enough to make routes work in cli projects) was not enough to fix the test.

gkalpak commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -191,6 +191,28 @@ export class NgtscProgram implements api.Program {
+      // not necessary, but keeping it backwards compatible in case someone else is using the API.
+
+      if (entryRoute.startsWith('.')) {
+        throw new Error('Resolution of relative paths is not supported.');
```
My initial intention was to keep this close to the old `AotCompilerHost` implementation, but I don't see any reason to :grin:
I'll add more info to the error.

gkalpak commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -191,6 +191,28 @@ export class NgtscProgram implements api.Program {
+      // Any containing file gives the same result for absolute entry path.
+      const containingFile = this.tsProgram.getRootFileNames()[0];
+      const [entryPath, moduleName] = entryRoute.split('#');
+      const resolved = ts.resolveModuleName(entryPath, containingFile, this.options, this.host);
```
But it has a comment that says: _"Any containing file gives the same result for absolute entry path."_ 😛

##
TBH, I find this whole `if` block pointless (it is just to keep the implementation backwards compatible with the old [AotCompilerHost][1]).
I'll try to comment it better.

[1]: https://github.com/angular/angular/blob/50732e156/packages/compiler-cli/src/transformers/compiler_host.ts#L175-L188

gkalpak commented on packages/compiler-cli/src/ngtsc/routing/src/route.ts:
```diff
@@ -15,18 +15,14 @@ export abstract class RouterEntryPoint {
 
   get name(): string { return this.moduleName; }
-
-  toString(): string { return `${this.filePath}#${this.moduleName}`; }
```
Happy to bring that back then :smiley:

gkalpak commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -2022,91 +2024,410 @@ describe('ngtsc behavioral tests', () => {
+        const routes = env.driveRoutes();
+        expect(routes).toEqual([
+          lazyRouteMatching(
+              './lazy#LazyModule', /\/test\.ts$/, 'TestModule', /\/lazy\.ts$/, 'LazyModule'),
```
Good point. I was just refactoring what was [there before][1].

[1]: https://github.com/angular/angular/pull/28542/files#diff-710a6b6957d5d5e7d5f0662d835e256eL2025

gkalpak commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -2022,91 +2024,410 @@ describe('ngtsc behavioral tests', () => {
+            selector: 'test-2',
+            template: '',
+          })
+          export class Test2Component {}
```
This is to ensure that having other things than `NgModule`s in the exports does not throw the RouteAnalyzer off.
I can split it out into a separate test.

gkalpak commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -2022,91 +2024,410 @@ describe('ngtsc behavioral tests', () => {
+        ]);
+      });
+
+      it('should ignore routes to unknown modules', () => {
```
Yeah, that's a good question. Not sure what it should do (but added a test for what it actually does right now).

I chose the current behavior (not to throw if it doesn't know about a lazy route's module), because that is somewhat similar to what would happen if `listLazyRoutes()` was called without an entry point: It would list all routes found in the know modules, but it would not complain if a lazy route points to an unknown module.

I can be convinced otherwise, though.

gkalpak committed:
fixup! refactor(ivy): remove unnecessary method from ngtsc's `RouterEntryPoint`

gkalpak committed:
fixup! test(ivy): fix `ngtools_api` test under ivy

gkalpak committed:
fixup! fix(ivy): support listing lazy route for project-root-relative entry point in `ngtsc`

gkalpak committed:
fixup! feat(ivy): implement listing lazy routes for specific entry point in `ngtsc`

gkalpak committed:
test(ivy): test listing lazy routes to different root directories

gkalpak commented:
Addressed the feedback in `fixup` and added a new commit with some tests with multiple `rootDirs` (to cover https://github.com/angular/angular/pull/28542#pullrequestreview-201288638).

Still need to decide https://github.com/angular/angular/pull/28542#discussion_r255029098 is what we want.

petebacondarwin reviewed (approved):
Great! I don't know what the answer to https://github.com/angular/angular/pull/28542#discussion_r255029098 is though.

alxhub commented:
[Presubmit](http://test/OCL:233473201:BASE:233473243:1549922306639:f50e5223)
